src/datahike/query.cljc

```diff
@@ -910,27 +929,39 @@
           stats? (assoc :tmp-stats {:type :rule
                                     :branches tmp-stats}))))))
 
+
+(defn resolve-pattern-lookup-entity-id [source e error-code]
+  (cond
+    (or (lookup-ref? e) (attr? e)) (dbu/entid-strict source e error-code)
+    (entid? e) e
+    (symbol? e) e
+    :else (or error-code (dt/raise "Invalid entid" {:error :entity-id/syntax :entity-id e}))))
+
 (defn resolve-pattern-lookup-refs
   "Translate pattern entries before using pattern for database search"
+  ([source pattern] (resolve-pattern-lookup-refs source pattern nil))
+  ([source pattern error-code]
+   (if (dbu/db? source)
+     (dt/with-elements-of pattern
+       e (resolve-pattern-lookup-entity-id source e error-code)
+       a (if (and (:attribute-refs? (dbi/-config source)) (keyword? a))
+           (dbi/-ref-for source a)
+           a)
+       v (if (and v (attr? a) (dbu/ref? source a) (or (lookup-ref? v) (attr? v)))
+           (dbu/entid-strict source v error-code)
+           v)
+       tx (if (lookup-ref? tx)
+            (dbu/entid-strict source tx error-code)
+            tx)
+       added added)
+     pattern)))
+
+(defn resolve-pattern-lookup-refs-or-nil
```

Comment:
Actually, it seems like I included this function for dealing with some of the test cases listed in this issue:

https://github.com/replikativ/datahike/issues/650

But because we haven't decided what to do in case a variable is bound to illegal values, maybe it is better to not deal with it. So I simply removed the function `resolve-pattern-lookup-refs-or-nil` and called `resolve-pattern-lookup-refs` instead. I also made other simplifications. This makes the PR more focused.

Reply:
OK, I realized I had to undo my previous change and put back the function `resolve-pattern-lookup-refs-or-nil`. I added an explanation in the docstring of the function why it is needed.